File: src/superviser.cpp
/*
 */

#include "superviser.h"
#include <sot/core/debug.hh>
#include <dynamic-graph/factory.h>
#include <dynamic-graph/all-commands.h>

#include <jrl/mal/matrixabstractlayer.hh>

using namespace dynamicgraph::sot;
using namespace dynamicgraph;

int find(const std::vector<std::string> &vec, const std::string &elmt)
{
  for(unsigned i=0;i<vec.size(); ++i)
    if(vec[i] == elmt)
      return i;
  return -1;
}

void disp(const std::vector<std::string> &vec, const std::string &name)
{
  std::cout << "  " << name << std::endl;
  for(unsigned i=0;i<vec.size(); ++i)
    std::cout << vec[i] << ", ";
  std::cout << std::endl;
}

void insert(std::vector<std::string> &vec, unsigned index, const std::string &elmt)
{
  std::vector<std::string>::iterator it = vec.begin();
  for(unsigned i=0; i < index; ++i)
    ++it;
  vec.insert(it,elmt);
}

void erase(std::vector<std::string> &vec, std::string elmt)
{
  for(std::vector<std::string>::iterator it = vec.begin(); it != vec.end(); )
  {
    if(*it == elmt)
    {
      vec.erase(it);
      it = vec.begin();
    }
    else
      ++it;
  }
}

DYNAMICGRAPH_FACTORY_ENTITY_PLUGIN(Superviser,"Superviser");

Superviser::
Superviser( const std::string & name )
  :Entity(name)
{
  sotDEBUGIN(5);
  initCommands();
  sotDEBUGOUT(5);
}


Superviser::
~Superviser( void )
{
}

void Superviser::clear()
{
  desiredStack_.clear();
}

void Superviser::push(const std::string & name)
{
  desiredStack_.push_back(name);
  return ;
}

void Superviser::update()
{
  std::vector<std::string> currentStack;
  unsigned i = 0;
  if(sot_ == 0x0)
  {
    std::cerr << "No solver bound, aborting" << std::endl;
    return;
  }

  while (sot_->getTask(i) != "")
  {
    currentStack.push_back(sot_->getTask(i));
    ++i;
  }

  if (currentStack == desiredStack_)
    return;

  //builds the list of operations required
  int index = 0;
  while (index < desiredStack_.size())
  {
    // identical task
    if (currentStack.size() > index && currentStack[index] == desiredStack_[index])
    {
      ++index;
    }
    //the task exists in b, but not at the same place
    else if (find(currentStack, desiredStack_[index]) != -1)
    {
      // one task has been removed
      if (currentStack.size() > index && find(desiredStack_, currentStack[index]) != -1)
      {
        sot_->removeByTaskName(currentStack[index]);
        erase(currentStack, currentStack[index]);
      }

      // the priority of the task a[index] has changed
      else
      {
        int index2 = find(currentStack, desiredStack_[index]);
        for(int i=index; i<index2; ++i)
          sot_->upByTaskName(desiredStack_[index]);
        erase(currentStack, desiredStack_[index]);
        insert(currentStack,index,desiredStack_[index]);
        ++index;
      }
    }
    // the task is not in b
    else
    {
      sot_->pushByTaskName(desiredStack_[index]);
      for(i=index; i<currentStack.size(); ++i)
      {
        sot_->upByTaskName(desiredStack_[index]);
      }
      insert(currentStack,index,desiredStack_[index]);
      ++index;
    }
  }
  
  // Now, we should have b = [a, some tasks], we remove the remaining tasks.
  while (index < currentStack.size())
  {
    sot_->removeByTaskName(currentStack[index]);
    erase(currentStack, currentStack[index]);
  }
}


void Superviser::display()
{
  if (sot_)
    std::cout << "+-SOT set    "     << std::endl;
  else
    std::cout << "+-SOT not set  "     << std::endl;
  for (unsigned i =0; i<desiredStack_.size(); ++i)
    std::cout << "| " << desiredStack_[i] <<std::endl;
}

void Superviser::
addSolverFromName( const std::string & solverName )
{
  sot_ = dynamic_cast< dg::sot::dyninv::SolverKine * > (&(dg::PoolStorage::getInstance()->getEntity(solverName)));
  if( sot_  == 0x0)
    std::cerr << "Unable to link with the solver" << std::endl;
}

void Superviser::initCommands()
{
  namespace dc = ::dynamicgraph::command;
  addCommand("clear",
    dc::makeCommandVoid0(*this,&Superviser::clear,
    "clear the list of desired task"));

  addCommand("push",
    dc::makeCommandVoid1(*this,&Superviser::push,
    "add a task at the list of desired stack"));

  addCommand("update",
    dc::makeCommandVoid0(*this,&Superviser::update,
    "replace the content of the stack by the desired list of task"));

  addCommand("disp",
    dc::makeCommandVoid0(*this,&Superviser::display,
    "display"));

  addCommand("setSoT",
    dc::makeCommandVoid1(*this,&Superviser::addSolverFromName,
    "set the solver"));
}

File: src/dynamic_graph/sot/robohow/superviser.py
#TODO: define the list of symbols exported.
from dynamic_graph import plug
from dynamic_graph.sot.core.operator import Norm_of_vector

"""
Publish the error of the associated task.
publisher: the rosPublish entity
taskname:  the task name 
"""
def startPublishingError(robot,publisher,taskname):
  if publisher == None:
    return

  # verify that the task exists in the robot database
  if taskname in robot.tasks and taskname != 'taskright-wrist':
    sig_name=taskname+'_error'
    sig_name = sig_name.replace('-', '_')
    
    # publishes the error
    publisher.add('vector', sig_name,'/sot/'+sig_name)
    plug(robot.tasks[taskname].error, publisher.signal(sig_name))

    # publishes the error norm
    sig_name=sig_name+'_norm'
    publisher.add('double', sig_name,'/sot/'+sig_name)
    errorNorm = Norm_of_vector('op_'+sig_name)
    plug(robot.tasks[taskname].error, errorNorm.sin)
    plug(errorNorm.sout, publisher.signal(sig_name))


"""
Stop the publication of the error, delete the corresponding signals
"""
def stopPublishingError(robot, publisher, taskname):
  if publisher == None:
    return

  if taskname in robot.tasks and taskname != 'taskright-wrist':
    # unplug the signal.
    sig_name=taskname+'_error'
    sig_name = sig_name.replace('-', '_')

    if publisher.hasSignal(sig_name):
      publisher.signal(sig_name).unplug()
      publisher.rm(sig_name)
    else:
      print "signal " + sig_name + " not found"

    sig_name=sig_name+'_norm'
    if publisher.hasSignal(sig_name):
      publisher.signal(sig_name).unplug()
      publisher.rm(sig_name)
    else:
      print "signal " + sig_name + " not found"





"""
The goal of the supervisor class is to ... 
"""
class Superviser:
  robot  = None
  solver = None
  publisher = None
  desiredStack = [] # desired state of the stack of tasks

  def __init__(self, robot, solver, publisher):
    self.robot = robot
    self.solver = solver
    self.publisher = publisher

  def getCurrentStack(self):
    taskList = self.solver.sot.getTaskList()
    currentStack = taskList.split('|')
    return currentStack

  """clear the list of desired task"""
  def clear(self):
    self.desiredStack = [];

  """add a task at the end of desired stack""";
  def push(self, name):
    self.desiredStack.append(name);


  """display the state of the current stack"""
  def dispCurrent(self):
    for name in self.getCurrentStack():
      print name

  """display the state of the desired stack"""
  def dispDesired(self):
    for name in self.desiredStack:
      print name

  """replace the content of the stack by the desired list of task"""
  def update(self):
    # builds the list of operations required
    index = 0
    currentStack = self.getCurrentStack()
    while (index < len(self.desiredStack)):
      # identical task
      if len(currentStack) > index and currentStack[index] == self.desiredStack[index]:
        index = index + 1
      # the task exists in b, but not at the same place
      elif self.desiredStack[index] in currentStack:
        # one task has been removed
        if len(currentStack) > index and not currentStack[index] in self.desiredStack:
          taskname = currentStack[index]
          stopPublishingError(self.robot, self.publisher, taskname)
          self.solver.sot.remove(taskname)
          currentStack.remove(taskname)

        # the priority of the task desiredStack_[index] has changed
        else:
          taskname = self.desiredStack[index]
          index2=currentStack.index(taskname)
          for i in range(index, index2):
            self.solver.sot.up(taskname)
          currentStack.remove(taskname)
          currentStack.insert(index,taskname)
          index = index + 1

      # the task is not in the real stack: add it
      else:
        taskname = self.desiredStack[index]
        self.solver.sot.push(taskname)
        startPublishingError(self.robot,self.publisher,taskname)

        for i in range(index, len(currentStack)):
          self.solver.sot.up(taskname)
        currentStack.insert(index,taskname)
        index = index+1
    
    # Now, we should have current = [desiredStack_, some tasks], we remove the remaining tasks.
    while (index < len(currentStack)):
      taskname = currentStack[index]
      stopPublishingError(self.robot, self.publisher, taskname)
      self.solver.sot.remove(taskname)
      currentStack.remove(taskname)

